### ISSUE TITLE
TC39 proposal: Array#flatten and Array#flatMap

### ISSUE THREAD

**Issue by jdalton** [underdash/underdash#17]
Brian Terlson (Microsoft :raised_hands: TC39 rep) has a [draft proposal](https://github.com/bterlson/proposal-flatMap) to add `Array#flatten` and `Array#flatMap`. While his `Array#flatMap` aligns [well-enough](https://github.com/bterlson/proposal-flatMap/issues/2) with Lodash's [_.flatMap](https://lodash.com/docs#flatMap) his `Array#flatten` is deep by default which is the opposite of Lodash's [_.flatten](https://lodash.com/docs#flatten).

Feedback on them would be appreciated. Let's keep spec discussion to [Brian's repo](https://github.com/bterlson/proposal-flatMap) and any Lodash/Underscore lib related side-bar discussions here.

_Update_
I've raised an issue there for [shallow by default](https://github.com/bterlson/proposal-flatMap/issues/9).

**Comment by jdalton** [comment]
I think `Array#flatten` should be shallow by default because I think it makes sense to do less work by default, it aligns to API's like the DOM `Node#cloneNode` which is shallow by default, and it would align with the existing ES3 pattern of using `Array#concat` for a shallow flatten.

**Comment by jdalton** [comment]
LOL I need to take my advice and move the spec discussion to Brian's repo. I raised an issue there for [shallow by default](https://github.com/bterlson/proposal-flatMap/issues/9)

**Comment by paulfalgout** [comment]
Assuming this does go through it seems like `_.flatten` would benefit from essentially equating to lodash's `_.flattenDepth` and the default can reflect the TC39 decision at some point.

**Comment by jdalton** [comment]
If the proposal isn't adjusted I think we're still fine as we would still be able to potentially use the built-in under the hood. I haven't come to terms with going deep by default yet and not sure the reasoning for it has been presented in a clear way at the moment (so time will tell).

**Comment by paulfalgout** [comment]
True, and I'm not necessarily suggesting if Infinity is the default value that underscore or lodash should switch.. just that if the TC39 spec establishes depth as an argument, `_.flatten` should probably accept that argument too.. then you really only need `_.flatten` and `_.flattenDepth` just becomes an alias
```js
_.flatten(myArray, 2) === _.flattenDepth(myArray, 2)
````
and less important..
```js
_.flatten(myArray, Infinity) === _.flattenDeep(myArray)
```

**Comment by jdalton** [comment]
Ah ok!

This is kind of a tricky spot. I've tried to split out shallow vs. deep to avoid [the boolean trap](http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html), compartmentalize functionality, and simplify composition.

This fit well as:

<table>
<tr>
<td>clone</td><td>cloneDeep</td>
</tr>
<tr>
<td>defaults</td><td>defaultsDeep</td>
</tr>
<tr>
<td>zipObject</td><td>zipObjectDeep</td>
</tr>
</table>

and internal helpers like
<table>
<tr>
<td>baseIsEqual</td><td>baseIsEqualDeep</td>
</tr>
<tr>
<td>baseMerge</td><td>baseMergeDeep</td>
</tr>
<tr>
<td>baseProperty</td><td>basePropertyDeep</td>
</tr>
</table>

I could see absorbing the `depth` param as counter to that.
The separation also affords the possibility of adding

* _.flatMapDeep
* _.flatMapDepth

which is a gap not covered by the draft proposal.

**Comment by jdalton** [comment]
I went ahead and added [_.flatMapDeep and _.flatMapDepth](https://github.com/lodash/lodash/commit/64704e16c14a0da7b7255cb4b729b8bbd46cf5ee).

**Comment by jdalton** [comment]
Looks like the `flatten` proposal is being updated to align with Lodash and be shallow by default.
